File: tests/resources/contracts/set_contract.ml
open Optitrust
open Prelude

let _ = Run.script_cpp (fun () ->
    !! Resources.(set_fun_contract [__modifies("a ~> Cell")]) [multi cFunDef ["incr"; "incr_twice"]];

    !! Resources.(set_fun_contract [__modifies("n ~> Cell"); __modifies("m ~> Cell")]) [cFunDef "incr_both"];

    (* Should produce empty diff: *)
    !! Resources.(set_fun_contract [__modifies("n ~> Cell, m ~> Cell")]) [cFunDef "incr_both"];

    !! Resources.(set_fun_contract [__modifies("m ~> Matrix1(sz)")]) [cFunDef "incr_range"];
    !! Resources.(set_loop_contract [__modifies("&m[MINDEX1(sz, i)] ~> Cell")]) [cFor "i"];

    (* Weird behavior because acc is already stack-var encoded when parsing the contract:
       we need to write `acc` and not `&acc` *)
    !! Resources.(set_loop_contract [__sequentially_modifies("acc ~> Cell")]) [cFunBody "loop"; cFor "j"];

    !! Resources.(set_loop_contract ~strict:false [__reads("&M[MINDEX1(n, j)] ~> Cell")]) [cFunBody "non_strict"; cFor "j"];
    (*!! Trace.apply Scope.unique_alpha_rename;*)

    Resources.show ();
)

File: tests/resources/contracts/set_contract.cpp
#include <optitrust.h>

void incr(int* a) {
  __admitted();
  *a = *a + 1;
}

void incr_twice(int* a) {
  incr(a);
  incr(a);
}

void incr_both(int* n, int* m) {
  incr(n);
  incr(m);
}

void incr_range(int* m, int sz) {
  for (int i = 0; i < sz; ++i) {
      incr(&m[MINDEX1(sz, i)]);
  }
}

void loop(int n) {
  __pure();
  int acc = 0;
  for (int j = 0; j < n; ++j) {
    ++acc;
  }
}

void non_strict(int* M, int n) {
  __reads("M ~> Matrix1(n)");
  int acc = 0;
  for (int j = 0; j < n; ++j) {
    acc += M[MINDEX1(n, j)];
  }
}
